diff --git a/svg_image.module b/svg_image.module
index 6e6682c..64342e0 100644
--- a/svg_image.module
+++ b/svg_image.module
@@ -9,6 +9,7 @@ use Drupal\file\Entity\File;
 use Drupal\svg_image\Plugin\Field\FieldFormatter\SvgImageFormatter;
 use Drupal\svg_image\Plugin\Field\FieldFormatter\SvgImageUrlFormatter;
 use Drupal\svg_image\Plugin\Field\FieldWidget\SvgImageWidget;
+use enshrined\svgSanitize\Sanitizer;
 
 /**
  * Implements hook_field_widget_info_alter().
@@ -108,3 +109,22 @@ function svg_image_get_image_file_dimensions(File $file) {
 function svg_image_is_file_svg(File $file) {
   return $file->getMimeType() === 'image/svg+xml';
 }
+
+/**
+ * Implements hook_file_insert().
+ *
+ * We implement this to make sure that an svg image gets sanitized when uploads.
+ */
+function svg_image_file_insert($file) {
+  if ($file->getMimeType() === 'image/svg+xml') {
+    $fileUri = $file->getFileUri();
+
+    if (file_exists($fileUri)) {
+      // Sanitize the SVG file to be safe.
+      $rawSvg = file_get_contents($fileUri);
+      $svgSanitizer = new Sanitizer();
+      $sanitized = ($svgSanitizer->sanitize($rawSvg));
+      file_put_contents($fileUri, $sanitized);
+    }
+  }
+}
